circuit/utils.go

```diff
@@ -55,3 +61,274 @@ func ComputeUserAssetsCommitment(api API, assets []UserAssetInfo) Variable {
 	commitment := poseidon.Poseidon(api, tmpUserAssets...)
 	return commitment
 }
+
+// one variable: TotalEquity + TotalDebt + BasePrice
+// one variable: VipLoanCollateral + MarginCollateral + PortfolioMarginCollateral
+// one variable contain two TierRatios and the length of TierRatios is even
+func GetVariableCountOfCexAsset(cexAsset CexAssetInfo) int {
+	res := 2 
+	res += len(cexAsset.VipLoanRatios) / 2
+	res += len(cexAsset.MarginRatios) / 2
+	res += len(cexAsset.PortfolioMarginRatios) / 2
+	return res
+}
+
+func ConvertTierRatiosToVariables(api API, ratios []TierRatio, res []Variable) {
+	for i := 0; i < len(ratios); i += 2 {
+		v := api.Add(ratios[i].Ratio, api.Mul(ratios[i].BoundaryValue, utils.Uint8MaxValueFr))
+		v1 := api.Add(api.Mul(ratios[i+1].Ratio, utils.Uint126MaxValueFr), api.Mul(ratios[i+1].BoundaryValue, utils.Uint134MaxValueFr))
+		res[i/2] = api.Add(v, v1)
+	}
+}
+
+func FillCexAssetCommitment(api API, asset CexAssetInfo, currentIndex int, commitments []Variable) {
+	counts := GetVariableCountOfCexAsset(asset)
+	commitments[currentIndex*counts] = api.Add(api.Mul(asset.TotalEquity, utils.Uint64MaxValueFrSquare),
+			api.Mul(asset.TotalDebt, utils.Uint64MaxValueFr), asset.BasePrice)
+	
+	commitments[currentIndex*counts+1] = api.Add(api.Mul(asset.VipLoanCollateral, utils.Uint64MaxValueFrSquare),
+			api.Mul(asset.MarginCollateral, utils.Uint64MaxValueFr), asset.PortfolioMarginCollateral)
+
+	ConvertTierRatiosToVariables(api, asset.VipLoanRatios, commitments[currentIndex*counts+2:])
+	ConvertTierRatiosToVariables(api, asset.MarginRatios, commitments[currentIndex*counts+2+len(asset.VipLoanRatios)/2:])
+	ConvertTierRatiosToVariables(api, asset.PortfolioMarginRatios, commitments[currentIndex*counts+2+len(asset.VipLoanRatios)/2+len(asset.MarginRatios)/2:])
+}
+
+func GenerateRapidArithmeticForCollateral(api API, r frontend.Rangechecker, tierRatios []TierRatio) {
+	tierRatios[0].PrecomputedValue = CheckAndGetIntegerDivisionRes(api, r, api.Mul(tierRatios[0].BoundaryValue, tierRatios[0].Ratio))
+	for i := 1; i < len(tierRatios); i++ {
+		api.AssertIsLessOrEqualNOp(tierRatios[i-1].BoundaryValue, tierRatios[i].BoundaryValue, 128, true)
+		api.AssertIsLessOrEqualNOp(tierRatios[i].Ratio, utils.PercentageMultiplierFr, 8, true)
+		api.AssertIsLessOrEqualNOp(tierRatios[i].BoundaryValue, utils.MaxTierBoundaryValueFr, 128, true)
+		diffBoundary := api.Sub(tierRatios[i].BoundaryValue, tierRatios[i-1].BoundaryValue)
+		current := CheckAndGetIntegerDivisionRes(api, r, api.Mul(diffBoundary, tierRatios[i].Ratio))
```

Comment:
Our boundary value is multiplied by 10^16, the reminder will be very small in real value. I think there is no edge case. If there is an edge case, Then in the witness generation, it will be detected.